### PR TITLE
docs(Contributing.md): Add Contributing.md

### PR DESCRIPTION
**Pull request by M-0-N-K**
docs ( Contributing.md): added Contributing.md

### CODE REVIEW

**Comment by chirgjn** [comment]
Let's use this convention instead:
https://www.conventionalcommits.org/en/v1.0.0/

Can you mention the steps to use the following CLI, so that commit messages are as per convention.
http://commitizen.github.io/cz-cli/

File: Contribution.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    6. Use whole words in names when possible.
+
+  ### Types
+    1. Do not export types/functions unless you need to share it across multiple components.
+    2. Do not introduce new types/values to the global namespace.
+    3. Shared types should be defined in `types.ts`.
+    4. Within a file, type definitions should come first.
+
+  ### `null` and `undefined`
+    1. Use `undefined`. Do not use null.
```
Comment (chirgjn) [inline Contribution.md]: What's the idea behind this?

File: Contribution.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
+# Contribution guidelines
+The guidelines for contribution to babitz backend project are listed below. Please go through it before any contribution.
+
+## 1. General guidelines
+- The general guidelines for writing code. We will be using a subset of [typescript contribution guidelines](https://github.com/microsoft/TypeScript/wiki/Coding-guidelines). In case any of these conflict with loopback4 guidelines, loopback4's guideline is taken as priority.
```
Comment (chirgjn) [inline Contribution.md]: Please link Loopback 4's guidelines.

File: Contribution.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
+# Contribution guidelines
+The guidelines for contribution to babitz backend project are listed below. Please go through it before any contribution.
+
+## 1. General guidelines
+- The general guidelines for writing code. We will be using a subset of [typescript contribution guidelines](https://github.com/microsoft/TypeScript/wiki/Coding-guidelines). In case any of these conflict with loopback4 guidelines, loopback4's guideline is taken as priority. 
+
+  ### Names
```
Comment (chirgjn) [inline Contribution.md]: Can you link to the respective sections of the TypeScript contribution guidelines.

File: Contribution.md [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        - `<T>(x: T, y: T) => x === y`
+    3. Always surround loop and conditional bodies with curly braces. Statements on the same line are allowed to omit braces.
+    4. Open curly braces always go on the same line as whatever necessitates them.
+    5. Parenthesized constructs should have no surrounding whitespace. <br />A single space follows commas, colons, and semicolons in those constructs. For example:
+        - `for (var i = 0, n = str.length; i < 10; i++) { }`
+        - `if (x < 10) { }`
+        - `function f(x: number, y: string): void { }`
+    6. Use a single declaration per variable statement <br />(i.e. use `var x = 1; var y = 2;` over `var x = 1, y = 2;`).
+    7. `else` goes on a separate line from the closing curly brace.
+    8. Use 1 tab per indentation.
```
Comment (chirgjn) [inline Contribution.md]: We are using `prettier` to take enforce style. Let's link the instruction of running the same here.

File: Contribution.md [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+## Loopback4 Guidelines
```
Comment (chirgjn) [inline Contribution.md:54]:
```suggestion
## File naming convention
```

**Review by chirgjn — CHANGES_REQUESTED** [review]

File: Contribution.md [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+- Limit the subject line to 50 characters
+- Capitalize the subject line
+- Wrap the body at 72 characters
+
+Reference: [how to write good commits](https://gist.github.com/qoomon/5dfcdf8eec66a051ecd85625518cfd13
+ https://chris.beams.io/posts/git-commit)
+
+## How to Raise Pull Requests 
+- Avoid making large pull requests so that code review gets easier.
+- Use either one feature/fix or one commit per pull request only.
```
Comment (chirgjn) [inline Contribution.md]: Please describe trunk based development.
Please add branch name examples.

File: Contribution.md [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+- Use ESLint to statically analyse the source code and detect common problems. Config used from loopback/eslint-config.
+- Prettier to keep our code always formatted the same way. Config stored in .prettierrc
+- Both of these can be done with `  npm run lint:fix  `
+
+
+## Testing
+
+- Write tests in the __test__ folder.
+- Make sure you cover at least 50% of the test-case scenario.
+- You can generate "test coverage" using the `npm run test:coverage` command.
```
Comment (chirgjn) [inline Contribution.md]: Add references for how to write tests and what tools are used for testing.

File: Contribution.md [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    7. `else` goes on a separate line from the closing curly brace.
+    8. Use 1 tab per indentation.
+
+## Loopback4 Guidelines
+- File Naming Convention taken from [angular's file naming convention](https://angular.io/guide/styleguide#separate-file-names-with-dots-and-dashes). Loopback internally uses this so this decision is taken to have consistency in file names. This states we use a file name to also represent what it does or is.
+    - an example is "user.model.ts" here we can see the file is a "model" and the name of the model is "user". 
+- We clearly define our api spec before coding the models and endpoints for them. Please have a look at how to [import api spec in loopback 4.](https://loopback.io/doc/en/lb4/OpenAPI-generator.html)
+
+
+## Directory structure
```
Comment (chirgjn) [inline Contribution.md]: Add relevant reference: https://loopback.io/doc/en/lb4/Loopback-application-layout.html

File: Contribution.md (hunk 1 above)
Comment (M-0-N-K) [reply Contribution.md]: Sometimes there are functions that accept parameters as null or object and although we could send a null in some variables where we control the flow, sometimes other code might send undefined or simply not send anything thus making the parameters undefined. This makes us do a check for both, param==null || param == undefined . We can avoid the extra check for null.

**Comment by chirgjn** [comment]
https://github.com/github/docs/blob/main/CONTRIBUTING.md

**Review by chirgjn — CHANGES_REQUESTED** [review]

**Comment by chirgjn** [comment]
@M-0-N-K a lot of the comments are still un-resolved

File: Contribution.md (hunk 2 above)
Comment (M-0-N-K) [reply Contribution.md]: done.

File: Contribution.md (hunk 3 above)
Comment (M-0-N-K) [reply Contribution.md]: done.

File: Contribution.md (hunk 4 above)
Comment (M-0-N-K) [reply Contribution.md]: done.

File: Contribution.md (hunk 6 above)
Comment (M-0-N-K) [reply Contribution.md]: done.

File: Contributing.md [hunk 9]
```diff
@@ -139,15 +140,17 @@ Commit messages should adhere to the following guidelines:
...
+Reference: [how to write good commits](https://www.conventionalcommits.org/en/v1.0.0/)
+Please use [commitizen](https://www.npmjs.com/package/commitizen) to format your commit messages.
+steps to use commitizen:
+- Install commitizen globally with, "npm install commitizen -g"
+- Setup commitizen's conventional-changelog adapter in your local setup with "commitizen init cz-conventional-changelog --yarn --dev --exact"
 
-## How to Raise Pull Requests 
-- Avoid making large pull requests so that code review gets easier.
+## How to Raise Pull Requests
+- We use trunk based developement for our developement. Please use the same.
```
Comment (chirgjn) [inline Contributing.md]: Please add a link for "trunk based development"

File: Contributing.md [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+- Separate subject from body with a blank line
+- Use the imperative mood in the subject line
+- Do not end the subject line with a period
+- Wrap the body at 72 characters
+
+Reference: [how to write good commits](https://www.conventionalcommits.org/en/v1.0.0/)
+Please use [commitizen](https://www.npmjs.com/package/commitizen) to format your commit messages.
+steps to use commitizen:
+- Install commitizen globally with, "npm install commitizen -g"
+- Setup commitizen's conventional-changelog adapter in your local setup with "commitizen init cz-conventional-changelog --yarn --dev --exact"
```
Comment (chirgjn) [inline Contributing.md]: We have already linked to the CLI documentation, hence this is redundant.
```suggestion
```

File: Contributing.md [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+
+## Linting and Formatting
+ Apart from the typescript [Style](https://github.com/microsoft/TypeScript/wiki/Coding-guidelines#style) we use prettier and ESLint.
+- Use ESLint to statically analyse the source code and detect common problems. Config used from loopback/eslint-config.
+- Prettier to keep our code always formatted the same way. Config stored in .prettierrc
+- Both of these can be done with `  npm run lint:fix  `
+- Instructions on [how to use prettier cli.](https://prettier.io/docs/en/cli.html)
+
+
+## [Testing](https://loopback.io/doc/en/lb4/Testing-your-application.html)
```
Comment (chirgjn) [inline Contributing.md:109]: Add a point linking to Loopback 4 testing best practices

File: Contributing.md [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+        - `<T>(x: T, y: T) => x === y`
+    3. Always surround loop and conditional bodies with curly braces. Statements on the same line are allowed to omit braces.
+    4. Open curly braces always go on the same line as whatever necessitates them.
+    5. Parenthesized constructs should have no surrounding whitespace. <br />A single space follows commas, colons, and semicolons in those constructs. For example:
+        - `for (var i = 0, n = str.length; i < 10; i++) { }`
+        - `if (x < 10) { }`
+        - `function f(x: number, y: string): void { }`
+    6. Use a single declaration per variable statement <br />(i.e. use `var x = 1; var y = 2;` over `var x = 1, y = 2;`).
+    7. `else` goes on a separate line from the closing curly brace.
+    8. Use 1 tab per indentation.
```
Comment (chirgjn) [inline Contributing.md]: We don't need these guidelines as we are using `prettier` for styling.
```suggestion
```

File: Contributing.md [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+```
+
+
+
+## Linting and Formatting
+ Apart from the typescript [Style](https://github.com/microsoft/TypeScript/wiki/Coding-guidelines#style) we use prettier and ESLint.
+- Use ESLint to statically analyse the source code and detect common problems. Config used from loopback/eslint-config.
+- Prettier to keep our code always formatted the same way. Config stored in .prettierrc
+- Both of these can be done with `  npm run lint:fix  `
+- Instructions on [how to use prettier cli.](https://prettier.io/docs/en/cli.html)
```
Comment (chirgjn) [inline Contributing.md]: Both prettier and ESLint will be installed as part of the project, let's only link to their home page and document the various "check" and "fix' commands here.

**Review by chirgjn — CHANGES_REQUESTED** [review]

File: Contributing.md [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+    4. Open curly braces always go on the same line as whatever necessitates them.
+    5. Parenthesized constructs should have no surrounding whitespace. <br />A single space follows commas, colons, and semicolons in those constructs. For example:
+        - `for (var i = 0, n = str.length; i < 10; i++) { }`
+        - `if (x < 10) { }`
+        - `function f(x: number, y: string): void { }`
+    6. Use a single declaration per variable statement <br />(i.e. use `var x = 1; var y = 2;` over `var x = 1, y = 2;`).
+    7. `else` goes on a separate line from the closing curly brace.
+    8. Use 1 tab per indentation.
+
+## [Loopback4 Guidelines](https://github.com/microsoft/TypeScript/wiki/Coding-guidelines)
```
Comment (chirgjn) [inline Contributing.md]: The link here is for TS guidelines

**Review by chirgjn — CHANGES_REQUESTED** [review]

File: Contribution.md (hunk 2 above)
Comment (chirgjn) [reply Contribution.md]: I can't find the link even in the latest commit.

File: Contributing.md (hunk 14 above)
Comment (chirgjn) [inline Contributing.md]: Let's rename this to "File naming convention"

**Review by chirgjn — CHANGES_REQUESTED** [review]

File: Contribution.md (hunk 8 above)
Comment (chirgjn) [reply Contribution.md]: This is still pending

File: Contributing.md [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+
+  ### [General Constructs](https://github.com/microsoft/TypeScript/wiki/Coding-guidelines#comments)
+    1. Do not use `for..in` statements; instead, use `ts.forEach`, `ts.forEachKey` and `ts.forEachValue`. Be aware of their slightly different semantics.
+    2. Try to use `ts.forEach`, `ts.map`, and `ts.filter` instead of loops when it is not strongly inconvenient.
+
+
+## File naming convention
+- File Naming Convention taken from [angular's file naming convention](https://angular.io/guide/styleguide#separate-file-names-with-dots-and-dashes). Loopback internally uses this so this decision is taken to have consistency in file names. This states we use a file name to also represent what it does or is.
+    - an example is "user.model.ts" here we can see the file is a "model" and the name of the model is "user".
+- We clearly define our api spec before coding the models and endpoints for them. Please have a look at how to [import api spec in loopback 4.](https://loopback.io/doc/en/lb4/OpenAPI-generator.html)
```
Comment (chirgjn) [inline Contributing.md:42]: Add a separate heading for API spec

**Review by chirgjn — CHANGES_REQUESTED** [review]

File: Contributing.md [hunk 16]
```diff
@@ -0,0 +1,135 @@
+# Contributing to this repository
+The guidelines for contribution to babitz backend project are listed below. Please go through it before any contribution.
+
+## General guidelines
+- The general guidelines for writing code. We will be using a subset of [typescript contribution guidelines](https://github.com/microsoft/TypeScript/wiki/Coding-guidelines). In case any of these conflict with loopback4 guidelines, loopback4's guideline is taken as priority.
```
Comment (chirgjn) [inline Contributing.md:5]: Please link to Loopback 4 guidelines

**Review by chirgjn — CHANGES_REQUESTED** [review]

**Comment by chirgjn** [comment]
@M-0-N-K *ping*

File: Contributing.md (hunk 16 above)
Comment (M-0-N-K) [reply Contributing.md:5]: done.

**Pull request merged**